<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Nuctools by homeveg</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Nuctools</h1>
      <h2 class="project-tagline">Software for analysis of chromatin feature occupancy profiles from high-throughput sequencing data</h2>
      <a href="https://github.com/homeveg/nuctools" class="btn">View on GitHub</a>
      <a href="https://github.com/homeveg/nuctools/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/homeveg/nuctools/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="nuctools" class="anchor" href="#nuctools" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>NucTools</h1>

<h2>
<a id="nuctools-is-a-software-package-for-analysis-of-chromatin-feature-occupancy-profiles-from-high-throughput-sequencing-data" class="anchor" href="#nuctools-is-a-software-package-for-analysis-of-chromatin-feature-occupancy-profiles-from-high-throughput-sequencing-data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>NucTools is a software package for analysis of chromatin feature occupancy profiles from high-throughput sequencing data</h2>

<p>Recent advancements in high-throughput sequencing methods create a vast amount of data in which numerous chromatin features are mapped along the genome. The results are frequently analysed by creating binarized data sets that link the presence/absence of a given readout to specific genomic loci. It is currently a challenge in the field to cope with continuous distributions of deep sequencing chromatin readouts, and to integrate the different types of datasets to reveal linkages between them. Here we introduce the NucTools suite of Perl scripts and a stand-alone visualization program for a nucleosome-centred downstream analysis of deep sequencing data. NucTools accounts for the continuous distribution of nucleosome occupancy. Furthermore, it is useful to associate nucleosome occupancy with other chromatin features like transcription factor binding, histone modifications or DNA methylation.</p>

<hr>

<h2>
<a id="system-requirement" class="anchor" href="#system-requirement" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SYSTEM REQUIREMENT</h2>

<p>Linux (2.6 kernel or later) or Mac (OSX 10.6 Snow Leopard or later) operating system
with minimal 16 GB of RAM is recommended*. Perl v5.8 or above is required.</p>

<p>The C/C++ compiling enviroment might be required for installing dependencies, such as bedtools. Systems may vary.
Please assure that your system has the essential software building packages (e.g. build-essential
for Fedora, XCODE for Mac...etc) installed properly before running the installing 
script.</p>

<p>NucTools was tested successfully on our Linux servers (CentOS release 6.7 w/ Perl v5.10.1; 
Fedora release 22 w/ Perl v5.20.3) and Macbook Pro laptops (MAC OSX 10.11 w/ XCODE v5.1).</p>

<p>*Memory requirments depends on the experimental system. For big genomes performance will increase greatly on machines with more memory. For example, processing of 22 mouse chromosomes, with the sequencing library size about 75 000 000 reads occupy at peak load around 60-70 Gb of RAM. Important to mention, the performance is very dependent on HDD read-write speed. Therefore the parallel running of many samples at once recomended only on the server-like system or computational clusters with RAID arrays, allowing real multithreaded read-write access to HDDs array.</p>

<hr>

<h2>
<a id="quick-start" class="anchor" href="#quick-start" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>QUICK START</h2>

<p>This is an example of profiling a "test.bed" file using NucTools. The testing BED file comes along with the NucTools package
in the "test" directory. More details are stated in the INSTRUCTION section.</p>

<ol>
<li><p>Obtaining NucTools package:</p>
<pre><code>$ git clone https://github.com/homeveg/nuctools.git NucTools</code></pre>
</li>
<li><p>Installing NucTools:
The NucTools package does not require installation. It is a collection of individual Perl scripts which can be executed individually.</p></li>
<li>
<p>Generate a genome annotation table using providing R script:</p>
<pre><code>$ Rscript misc/LoadAnnotation.BioMart.R</code></pre>
</li>
<li>
<p>Prepare BED files from BAM files with external application (optionally) </p>
<ol>
<li>
<p>merge multiple replicates to one BAM file and sort by reads name:  </p>

<pre><code>$ samtools merge -n /Path_to_folder_with/BAM/test_sorted.bam /Path_to_folder_with/BAM/test.rp1.bam /Path_to_folder_with/BAM/test.rp2.bam /Path_to_folder_with/BAM/test.rp3.bam</code></pre>
</li>
<li>
<p>converted sorted BAM file to BED using bowtie2bed.pl script or with external app bedtools:  </p>
<pre><code>$ perl -w bowtie2bed.pl -i /Path_to_folder_with/BAM/test_sorted.bam -verbose > /Path_to_folder_with/BED/test_sorted.bed.gz
# using bedtools
$ bedtools bamtobed -i /Path_to_folder_with/BAM/test_sorted.bam | pigz &gt; /Path_to_folder_with/BED/test_sorted.bed.gz
</code></pre>
</li>
</ol>
</li>
<li>
<p>Running NucTools: </p>

<ol>
<li>
<p>Extend single-end reads to the average DNA fragment size (100):  </p>
<pre><code>$ extend_SE_reads.pl -in test.bed -out test.ext.bed.gz</code></pre>
</li>
<li>
<p>Extract ndividual chromosomes from whole genome BED:  </p>
<pre><code>$ extract_chr_bed.pl -in test.ext.bed.gz -out test -d /Path_to_folder_with/BED/ -p chr</code></pre>
</li>
<li>
<p>Convert all BED files to occupancy OCC files avereging nucleosomes occupancy values over the window of size 10:  </p>
<pre><code>$ bed2occupancy_average.pl -in /Path_to_folder_with/BED/ -odir /Path_to_folder_with/OCC -dir -use -w 10</code></pre>
</li>
<li>
<p>Calculate aggregate profiles and aligned occupancy matrices for each chromosome individually:  </p>
<pre><code>$ aggregate_profile.pl -reg genome_annotation.tab -idC 0 -chrC 4 --strC 7 -sC 8 -eC 9 -pbN -lsN -lS  -chr 1 -al /Path_to_folder_with/OCC/chr1.test.occ_matrix -av /Path_to_folder_with/OCC/chr1.test.aggregate -in /Path_to_folder_with/OCC/chr1.test.w10.occ.gz -upD 1000 -downD 1000</code></pre>
</li>
<li>
<p>Paste together aggregate profiles of each chromosome in one file and add a header</p>
<ul>
<li>Generate a header:<br>
</li>
</ul>
<pre><code>$ ls /Path_to_folder_with/OCC/<em>1000_1000.txt | perl -n -e 'if(/.</em>(chr.<em>).test.</em>/gm) { print $1, "\t"; }' | perl -n -e 'if( /(.*)\t$/g )  { print $1}' &gt; /Path_to_folder_with/OCC/test.all.occ.txt
$ echo "" &gt;&gt; /Path_to_folder_with/OCC/test.all.occ.txt</code></pre>
<ul>
<li>paste all aggregate profiles to one file<br>
</li>
</ul>
<pre><code>$ paste /Path_to_folder_with/OCC/*1000_1000.txt &gt;&gt; /Path_to_folder_with/OCC/test.all.occ.txt</code></pre>
</li>
</ol>
</li>
<li><p>Optionally: 
Visualize aggregate profiles and run K-mean cluster analysis on aligned occupancy matrixes with the MatLab-based ClusterMaps Building Tool (provided as a part of NucTools package). 
<a href="https://github.com/homeveg/nuctools/raw/master/CMB/ClusterMaps_Builder_src.zip">Download link</a></p></li>
</ol>

<hr>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>the NucTools suite for a nucleosome-centred downstream analysis of deep sequencing data is primarily Perl-based, and require at least Perl v5.8 with dependencies installed properly (listed in README_FULL.md).
A visualisation program is written on MatLab and requires either full MatLab insatllation or can be provided as a standolone application with web-installer compiled for MacOS X 10.9 or later.
NucTools utilize whole genome <a href="https://genome.ucsc.edu/FAQ/FAQformat#format1">BED</a> files.</p>

<p><strong>Optional external applications:</strong></p>

<ul>
<li>
<a href="http://samtools.sourceforge.net/">SamTools</a> - merge and sort BAM files</li>
<li>
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a> - convert BAM to BED</li>
<li>
<a href="http://zlib.net/pigz/">PIGZ</a> - a parallel implementation of gzip for modern multi-processor, multi-core machines</li>
</ul>

<hr>

<h3>
<a id="running-nuctools" class="anchor" href="#running-nuctools" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Running NucTools</h3>

<p>One can devide the NucTools pipeline to 3 major steps: (1) prepare input OCC files (2) calculate aggregate profiles and alligned occupancy matrixes (3) followup analysis and results visualization. 
For the moment we don't have a wrapper to run all 3 steps automatically so, each step should be executed separately and, in turn, consists of several intermediate steps.</p>

<p>We will provide a testing BAM file, command line bash file, and example output in ./test. Below is a set of commands that runs "test.bam" through NucTools pipeline:</p>

<pre><code>    $ samtools sort -n ./test/test_sorted.bam ./test/test.bam
    $ bedtools bamtobed -i ./test/test_sorted.bam | pigz &gt; ./test/test_sorted.bed.gz
    $ extend_SE_reads.pl -in ./test/test.bed -out ./test/test.ext.bed.gz
    $ extract_chr_bed.pl -in ./test/test.ext.bed.gz -out test/BED -d ./test -p chr 
    $ bed2occupancy_average.pl -in ./test/BED -odir ./test/OCC -dir -use -w 10
    $ aggregate_profile.pl -reg genome_annotation.txt -idC 0 -chrC 4 -strC 7 -sC 8 -eC 9 -pbN -lsN -lS 75000000 -chr 1 -al ./test/OCC/chr1.test.occ_matrix -av ./test/OCC/chr1.test.aggregate -in ./test/OCC/chr1.test.w10.occ.gz -upD 1000 -downD 1000
</code></pre>

<p>In the example above test.bam file is sorted by the reads names and converted to test_sorted.bed.gz file. As long as in the example case we are dealing with single-end ilumina sequencing reads, with expected read length of 100, we extending each read to the length of 100 using extend_SE_reads.pl. Resulting whole genome BED file is splitted to chromosomes with extract_chr_bed.pl script and all per-chromosome bed files are converted to OCC files with bed2occupancy_average.pl, using running window 10 (use -w 0 for bep-base resolution). Last step is to generate an aggregate profiles using aggregate_profile.pl script </p>

<hr>

<h3>
<a id="interpreting-results" class="anchor" href="#interpreting-results" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Interpreting Results</h3>

<p>NucTools bed2occupancy_average.pl scritp generates 2 types of the output. The file containing one column of numbers, corresponding to an aggregate profile (could be visualized, for example in Excel) and the tab-delimited text file containing all occupancy data for each region of interest (or transcript), aligned at annotated strat. These matrixes can be later visualized with our ClusterMaps Building Tool.</p>

<hr>

<h2>
<a id="additional-information" class="anchor" href="#additional-information" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Additional information</h2>

<p>Additional information, publications references and short description of each script from the toolbox can be found here:</p>

<ul>
<li>
<a href="http://www.generegulation.info/index.php/nuctools">http://www.generegulation.info/index.php/nuctools</a> (external link)</li>
<li>
<a href="https://homeveg.github.io/nuctools/">https://homeveg.github.io/nuctools/</a> (GitHub pages)</li>
</ul>

<p><strong>PLEASE NOTE</strong>:
<em>we are currently preparing manuscript and will add here more information, usage instruction and test data set soon.
The project is under development now.</em></p>

<h3>
<a id="planned-modifications" class="anchor" href="#planned-modifications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Planned modifications</h3>

<p>There are several changes has been planned and partially implemented:</p>

<ul>
<li>implement GZIP support for all input/output files generated by the pipeline scripts</li>
<li>implement gapped OCC format support to all relevant scripts</li>
<li>extend NucTools documentation and usage examples with Perl POD package</li>
<li>unify pipeline scripts input parameters using Perl GETOPT package</li>
</ul>

<h3>
<a id="future-possible-modifications" class="anchor" href="#future-possible-modifications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Future possible modifications</h3>

<ul>
<li>BAM files support</li>
<li>parallel processing (beautiful codes snippets for implementation of parallel processing of BAM files with Perl one can find here: <a href="https://genomebytes.wordpress.com/2013/07/24/multi-thread-access-of-bam-files-using-perl-and-samtools/">https://genomebytes.wordpress.com/2013/07/24/multi-thread-access-of-bam-files-using-perl-and-samtools/</a> )</li>
<li>NucTools automated package installation with make</li>
</ul>

<h3>
<a id="developers" class="anchor" href="#developers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Developers:</h3>

<p>Yevhen Vainshtein and Vladimir B. Teif</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/homeveg/nuctools">Nuctools</a> is maintained by <a href="https://github.com/homeveg">homeveg</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
Contact GitHub API Training Shop Blog About
© 2016 GitHub, Inc. Terms Privacy Security Status Help
