<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Nuctools by homeveg</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Nuctools</h1>
      <h2 class="project-tagline">Software for analysis of chromatin feature occupancy profiles from high-throughput sequencing data</h2>
      <a href="https://github.com/homeveg/nuctools" class="btn">View on GitHub</a>
      <a href="https://github.com/homeveg/nuctools/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/homeveg/nuctools/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="nuctools" class="anchor" href="#nuctools" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>NucTools</h1>

<h2>
<a id="nuctools-is-a-software-package-for-analysis-of-chromatin-feature-occupancy-profiles-from-high-throughput-sequencing-data" class="anchor" href="#nuctools-is-a-software-package-for-analysis-of-chromatin-feature-occupancy-profiles-from-high-throughput-sequencing-data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>NucTools is a software package for analysis of chromatin feature occupancy profiles from high-throughput sequencing data</h2>

<p>Recent advancements in high-throughput sequencing methods create a vast amount of data in which numerous chromatin features are mapped along the genome. The results are frequently analysed by creating binarized data sets that link the presence/absence of a given readout to specific genomic loci. It is currently a challenge in the field to cope with continuous distributions of deep sequencing chromatin readouts, and to integrate the different types of datasets to reveal linkages between them. Here we introduce the NucTools suite of Perl scripts and a stand-alone visualisation program for a nucleosome-centred downstream analysis of deep sequencing data. NucTools accounts for the continuous distribution of nucleosome occupancy. Furthermore, it is useful to associate nucleosome occupancy with other chromatin features like transcription factor binding, histone modifications or DNA methylation.</p>

<h2>
<a id="additional-information" class="anchor" href="#additional-information" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Additional information</h2>

<p>Additional information, publications references and short description of each script from the toolbox can be found here:
<a href="http://www.generegulation.info/index.php/nuctools">http://www.generegulation.info/index.php/nuctools</a></p>

<p><strong>PLEASE NOTE:</strong>
<em>we are currently preparing manuscript and will add here more information, usage instruction and test data set soon.
The project is under development now.</em></p>

<h4>
<a id="planned-modifications" class="anchor" href="#planned-modifications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Planned modifications:</h4>

<p>There are several changes has been planned and partially implemented:</p>

<ul>
<li>implement GZIP support for all input/output files generated by the pipeline scripts </li>
<li>implement gapped OCC format support to all relevant scripts</li>
<li>extend NucTools documentation and usage examples with Perl POD package</li>
<li>unify pipeline scripts input parameters using Perl GETOPT package</li>
</ul>

<h4>
<a id="future-possible-modifications" class="anchor" href="#future-possible-modifications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Future possible modifications:</h4>

<ul>
<li>BAM files support</li>
<li>parallel processing (beautiful codes snippets for implementation of parallel processing of BAM files with Perl one can find here: <a href="https://genomebytes.wordpress.com/2013/07/24/multi-thread-access-of-bam-files-using-perl-and-samtools/">https://genomebytes.wordpress.com/2013/07/24/multi-thread-access-of-bam-files-using-perl-and-samtools/</a> )</li>
<li>NucTools automated package installation with make </li>
</ul>

<h3>
<a id="developers" class="anchor" href="#developers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Developers:</h3>

<p>Yevhen Vainshtein and Vladimir B. Teif</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/homeveg/nuctools">Nuctools</a> is maintained by <a href="https://github.com/homeveg">homeveg</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
